### ISSUE TITLE
Application performance differences using mpirun/mpiexec with v3.1.1

### ISSUE THREAD

**Issue by stefan-vaisala** [aws/aws-parallelcluster#3788]
Hello!

We are noticing some significant performance differences when running our MPI application with PCluster version 3.1.1. These differences stem from using mpirun/mpiexec via the Hydra bootstrap process manager compared to using PMI2 and srun from Slurm for job submission. Mainly, the job is extremely slow and unbalanced when using mpirun/mpiexec while using srun provides expected performance. This was not an issue with PCluster version 3.0.1. Specifics are:

- AWS ParallelCluster version [e.g. 2.9.0]: ParallelCluster 3.1.1
- Intel oneAPI 2021.4 (Intel MPI 2021.4)
- AWS OFI/Libfabric with EFA, v 1.14.1

We currently run our MPI/openMP application by setting the following variables to enable the use of the Hydra bootstrap and mpirun/mpiexec:
```
export I_MPI_OFI_LIBRARY_INTERNAL=0
export I_MPI_PIN_RESPECT_CPUSET=0
export I_MPI_HYDRA_BOOTSTRAP=slurm
export I_MPI_HYDRA_RMK=slurm
```

When running the job with the above on a cluster created by PCluster v3.1.1, the job does not seem to be properly distributed and runs extremely slow (several 100% slower). The same exact job submission and performance on a cluster created by PCluster v3.0.1 exhibits normal application performance.

If we set the following to use Slurm srun with PMI2 and disabling the use of Hydra bootstrap:
```
export I_MPI_OFI_LIBRARY_INTERNAL=0
export I_MPI_PMI_LIBRARY=/opt/slurm/lib/libpmi2.so
export SLURM_MPI_TYPE=pmi2
```

The application runs with the expected performance on a cluster created with PCluster v3.1.1.

Were there any changes to the underlying AMI that could cause such a discrepancy? Under the hood, the Hydra process manager leverages srun to run the job, so its interesting that there is a difference between the two methods. I know Slurm was upgraded in v3.1.1 as well, so I am curious if there was an unforeseen change to task affinity/distribution (or Intel handling task distribution to/from Slurm is broken...again). Further - using srun and PMI is the preferred method by Slurm, and one we have used before. So - is there a recommendation from PCluster on which job execution method to use? We'd like to use the one that is supported/tested the most as to not encounter version to version runtime differences. 

Thanks!

**Comment by demartinofra** [comment]
Hi @stefan-maxar 

Thanks for reporting this. As you mentioned ParallelCluster 3.1.1 introduced the new Slurm 21.08 and the new Intel MPI 2021.4. Slurm configuration remained pretty much the same so the difference you are seeing should be related to changes in IntelMPI and/or Slurm. 

Our recommendation is to always follow Slurm best practices when submitting MPI jobs with Slurm in ParallelCluster: https://slurm.schedmd.com/mpi_guide.html

At the same time I'd like to reproduce this problem on our side in order to understand if there is any issue with any of these new versions. Could you share with us how are you submitting the application both in terms of Slurm options and mpi command?Also could you set `export I_MPI_DEBUG=10` and check from the debug output if node allocation and CPU reservation is correct?

**Comment by stefan-vaisala** [comment]
Hey @demartinofra 

Thanks for the quick feedback and the recommendation! We have used both job submission methods (Hydra process manager or direct PMI2 integration) with PCluster + Slurm and this is the first time we've noticed such a stark discrepancy. 

Given that the srun-PMI2 integration (https://slurm.schedmd.com/mpi_guide.html#intel_srun) seems to be more tightly coupled, we will use that going forward. Only concern I have with that is if PMI2 libraries are no longer supported at sometime in the future, as Intel MPI does not currently support PMIx (see: https://www.intel.com/content/www/us/en/develop/documentation/mpi-developer-guide-linux/top/running-applications/job-schedulers-support.html). PMIx _should_ work...but I havent tested it exhaustively enough to say one way or the other. Will cross that bridge when we get there.

Job submission is as follows. The source and export statements are actually in our .bashrc file, but for clarity and copy and paste, putting them directly here.
```
export I_MPI_OFI_LIBRARY_INTERNAL=0
source /shared/intel/setvars.sh # our full version of Intel oneAPI 2021.4 (Base + HPC kits)
export LD_LIBRARY_PATH=/opt/amazon/efa/lib64:$LD_LIBRARY_PATH
export PATH=/opt/amazon/efa/bin:$PATH
export I_MPI_PIN_RESPECT_CPUSET=0
export I_MPI_HYDRA_BOOTSTRAP=slurm
export I_MPI_HYDRA_RMK=slurm
export I_MPI_PIN=1
# Job submission:
sbatch -p [queue] -J [jobname] -n [procs] -c [OMP threads] -N [number of nodes] -o slurm-%x.out run_script.sh
# where: run_script.sh contains:
#   export OMP_NUM_THREADS=[OMP threads]
#   mpiexec [mpi application]
```

I set I_MPI_DEBUG=5 for some initial testing and didnt see anything totally out of the ordinary, but will try again with higher debug. What was evident is that even with the MPI and OMP settings requested via sbatch, only a small fraction of the cores per compute node were being utilized by the job even though the job submitted the correct number of tasks to the compute node. That is, using top/htop, I could see only a very small segment of cores being utilized when all cores of the compute node should have been utilized based on the job submitted. My hunch is something is up with the translation/distribution of OMP tasks...but havent dug that far in yet.

**Comment by demartinofra** [comment]
I'm able to reproduce on my side and the problem seems to be related to how processes are distributed. In fact some processes end up being pinned to the same core, therefore the performance degradation. When using --ntasks-per-core it looks like I can better control the process distribution, however I do see a different behavior between Slurm 21.08 and Slurm 20.11.

i'll engage SchedMD to better understand if this change in behavior is expected

**Comment by stefan-vaisala** [comment]
Glad you were able to reproduce! Yes, that sounds exactly like the issue and would make sense why I was seeing the correct number of tasks but only a fraction of available cores being used - the tasks were all binding to the same set of cores. Why this is impacting Hydra process manager - Slurm integration and not use of PMI2, not entirely sure. The following works just fine and distributes the tasks as expected:
```
export I_MPI_OFI_LIBRARY_INTERNAL=0
source /shared/intel/setvars.sh # our full version of Intel oneAPI 2021.4 (Base + HPC kits)
export LD_LIBRARY_PATH=/opt/amazon/efa/lib64:$LD_LIBRARY_PATH
export PATH=/opt/amazon/efa/bin:$PATH
export I_MPI_PMI_LIBRARY=/opt/slurm/lib/libpmi2.so
export SLURM_MPI_TYPE=pmi2
export I_MPI_PIN=1
# Job submission:
sbatch -p [queue] -J [jobname] -n [procs] -c [OMP threads] -N [number of nodes] -o slurm-%x.out run_script.sh
# where: run_script.sh contains:
#   export OMP_NUM_THREADS=[OMP threads]
#   mpiexec [mpi application]
```

**Comment by demartinofra** [comment]
According the SchedMD the issue seems related to https://bugs.schedmd.com/show_bug.cgi?id=13351.

**Comment by stefan-vaisala** [comment]
> According the SchedMD the issue seems related to https://bugs.schedmd.com/show_bug.cgi?id=13351.

Awesome - thanks for finding that. Will transition to using `srun` and the Slurm-PMI2 interface per SchedMD's recommendation with PCluster v3.1.1. 

Side note:
I did try to use the PMI2 and PMIx libraries from PMIx (found in: `/opt/pmix/lib/`) by setting `export I_MPI_PMI_LIBRARY=/opt/pmix/lib/['libpmi2.so or libpmix.so]` and it does not work with Intel MPI 2021.4. The Slurm-based PMI2 lib (`/opt/slurm/lib/libpmi2.so`) does work. Not if this is expected behavior or not and based on how Slurm is configured and compiled.

**Comment by hanwen-pcluste** [comment]
Parallelcluster 3.1.2 has been released, containing Slurm 21.08.6 with the fix
I am going to close this issue. Please let us know if you have further concerns.

Thank you!

**Comment by stefan-vaisala** [comment]
Confirmed PCluster 3.1.2 with Slurm 21.08.6 fixes the issue. Thanks!
